### PR TITLE
[MPDX-8349] Add monthly commitment average and goal

### PR DESCRIPTION
**Pull request by canac**
## Description

The previous monthly commitment average and goal was inaccurate compared to Angular. This PR adds the correct method for calculating the average and goal. It also moves the complex logic into helper functions and hooks.

[MPDX-8349](https://jira.cru.org/browse/MPDX-8349)

## Checklist:

- [x] I have given my PR a title with the format "MPDX-(JIRA#) (summary sentence max 80 chars)"
- [x] I have applied the appropriate labels. (_Add the label "On Staging" to get the branch automatically merged into staging._)
- [x] I have requested a review from another person on the project

### CODE REVIEW

File: src/components/Coaching/CoachingDetail/MonthlyCommitment/MonthlyCommitment.test.tsx [hunk 1]
```diff
@@ -1,18 +1,64 @@
...
+                pledged: i * 10,
+              },
+        ),
+      },
+      MonthlyCommitmentSingleMonth: {
+        reportPledgeHistories: jest
+          .fn()
+          .mockReturnValueOnce([{ pledged: 50, received: 50 }])
+          .mockReturnValueOnce([{ pledged: 100, received: 150 }]),
+      },
```
Comment (dr-bizz) [inline src/components/Coaching/CoachingDetail/MonthlyCommitment/MonthlyCommitment.test.tsx:50]: This is a great way to call it twice and get different results.

File: src/components/Coaching/CoachingDetail/MonthlyCommitment/useMonthlyCommitmentAverage.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    if (!mpdInfo?.activeMpdStartAt) {
+      return null;
+    }
+
+    const startAt = DateTime.fromISO(mpdInfo.activeMpdStartAt);
+    if (startAt > DateTime.now()) {
+      // If the start at date is in the future, don't load commitment data for the start at date
+      return null;
+    }
+    return startAt.endOf('month');
```
Comment (dr-bizz) [inline src/components/Coaching/CoachingDetail/MonthlyCommitment/useMonthlyCommitmentAverage.ts:34]: What if the end of the month is in the future?

File: src/components/Coaching/CoachingDetail/MonthlyCommitment/useMonthlyCommitmentAverage.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+  // Calculate the last month that the user was on active MPD
+  const finishMonth = useMemo(() => {
+    if (!mpdInfo?.activeMpdFinishAt) {
+      return null;
+    }
+
+    const finishAt = DateTime.fromISO(mpdInfo.activeMpdFinishAt);
+    // If the finish at date is in the future, clamp it to the end of the current month
+    return DateTime.min(finishAt, DateTime.now()).endOf('month');
```
Comment (dr-bizz) [inline src/components/Coaching/CoachingDetail/MonthlyCommitment/useMonthlyCommitmentAverage.ts:45]: It looks like you are always clamping it to the end of the month, which could not be in the future. Maybe I'm understanding the comment incorrectly.

File: src/components/Coaching/CoachingDetail/MonthlyCommitment/useMonthlyCommitmentAverage.ts
Comment (dr-bizz) [inline src/components/Coaching/CoachingDetail/MonthlyCommitment/useMonthlyCommitmentAverage.ts:1]: This page is awesome

**Review by dr-bizz — APPROVED** [review]
This is  an awesome update! great coding!

File: pages/api/graphql-rest.page.ts [hunk 4]
```diff
@@ -447,7 +453,7 @@ class MpdxRestApi extends RESTDataSource {
   ) {
     const designationAccountFilter =
       designationAccountId && designationAccountId.length > 0
-        ? `&filter[designation_account_id=${designationAccountId.join(',')}`
+        ? `&filter[designation_account_id]=${designationAccountId.join(',')}`
```
[the remark this answers is not in the document]
Comment (canac) [reply pages/api/graphql-rest.page.ts:456]: I didn't bother testing it, but I'm pretty sure that designation account id filter wouldn't work how it was before.

File: src/components/Coaching/CoachingDetail/MonthlyCommitment/useMonthlyCommitmentAverage.ts (hunk 3 above)
Comment (canac) [reply src/components/Coaching/CoachingDetail/MonthlyCommitment/useMonthlyCommitmentAverage.ts:45]: The goal is to use the end of this month or the end of the finish at month, whichever comes first. It's OK if the finish at month is in the past. But if it's in the future, we want to use the end of the current month to not mess up the average calculation. Do you think the comment could be improved?

File: src/components/Coaching/CoachingDetail/MonthlyCommitment/useMonthlyCommitmentAverage.ts (hunk 2 above)
Comment (canac) [reply src/components/Coaching/CoachingDetail/MonthlyCommitment/useMonthlyCommitmentAverage.ts:34]: If the end month is in the future, it gets clamped to the end of the current month.

File: src/components/Coaching/CoachingDetail/MonthlyCommitment/useMonthlyCommitmentAverage.ts (hunk 3 above)
Comment (dr-bizz) [reply src/components/Coaching/CoachingDetail/MonthlyCommitment/useMonthlyCommitmentAverage.ts:45]: your comment makes sense, I think I was misunderstanding.

**Pull request merged**